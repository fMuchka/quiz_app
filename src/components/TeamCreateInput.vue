<template>
  <div class="input-wrapper">
    <label for="team-input" :style="textColor">{{ color.label }} tým</label>

    <input
      name="team-input"
      v-model="label"
      :readonly="isInputDisabled"
      type="text"
    />

    <button name="confirm-button" @click="confirmTeam()" :disabled="isConfirmDisabled">✓</button>

    <button name="delete-button" @click="$emit('remove')">✗</button>
  </div>
</template>

<script>
export default {
  name: "TeamCreateInput",

  data() {
    return {
      label: '',
      inputDisabled: false
    };
  },

  computed: {
    color: {
      get() {
        return this.$props.team.color;
      },

      set() {},
    },
    textColor() {
        return "color:" + this.$props.team.color.id;
    },
    isConfirmDisabled(){
        return (this.label.length > 0) ? false : true;
    },
    isInputDisabled: {
      get(){
        return this.inputDisabled;
      },
      set(status){
        this.inputDisabled = status;
      }
    }
  },
  props: {
    index: Number,
    team: Object,
  },

  methods: {
    confirmTeam() {
      const buttons = [
        this.$el.children[this.$el.children.length - 1],
        this.$el.children[this.$el.children.length - 2],
      ];

      buttons.forEach((el) => {
        el.remove();
      });

      const team = {
        label: this.label,
        color: this.color.id,
      };

      this.$store.commit("pushTeam", team);
      this.isInputDisabled = true;
    },
  },
};
</script>

<style>
.input-wrapper {
  display: grid;
  grid-template-columns: 200px 250px 60px 60px;
  align-items: center;
  margin-bottom: 0.5rem;
}

.input-wrapper * {
  display: grid;
  font-size: 30px;
}

#pageWrapper * button[name="confirm-button"] {
  margin: 0 0px 0 10px;
}

#pageWrapper * button {
  margin: 0 5px;
}

button[name="confirm-button"] {
  background-color: var(--secondary-color);
  color: var(--main-color);
  border: none;
}

button[name="confirm-button"]:hover,
button[name="delete-button"]:hover {
  cursor: pointer;
  color: 003049;
}

button[name="delete-button"] {
  background-color: #e00000;
  color: var(--main-color);
  border: none;
}
</style>