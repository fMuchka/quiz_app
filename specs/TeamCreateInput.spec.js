import Vuex from "vuex";
import { mount, createLocalVue } from "@vue/test-utils";

import TeamCreateInput from "../src/components/TeamCreateInput.vue"

const localVue = createLocalVue();
localVue.use(Vuex);

describe("TeamCreateInput Component:", () => {
    let store;
    let props;

    beforeEach(() => {
      store = new Vuex.Store({
        state: {
          teams: [],
          colorList: [
            { id: 'red', label: 'Červený', used: false },
            { id: 'blue', label: 'Modrý', used: false },
            { id: 'green', label: 'Zelený', used: false },
            { id: 'yellow', label: 'Žlutý', used: false },
            { id: 'purple', label: 'Fialový', used: false },
            { id: 'brown', label: 'Hnědý', used: false },
          ]
        },
        mutations: {
          pushTeam(state, team) {
            state.teams.push(team);
          }
        },
        getters: {
          color: (state) => {
            const colors = state.colorList;
            const freeIndex = store.getters.nextColorIndex;

            if (freeIndex === false) {
              // no colors left
              return false;
            }

            if (freeIndex < colors.length) {
              return colors[freeIndex];
            }
          }
        }
      });
        
        props = {
            index: 0,
            team: {
                id: 0,
                color: store.state.colorList[0]
            }
        };
    });

    it('matches snapshot', () => {
        const wrapper = mount(TeamCreateInput, {
            localVue,
            store,
            propsData:  props
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it('creates with correct properties', () => {
        const wrapper = mount(TeamCreateInput, {
            localVue,
            store,
            propsData:  props
        });

        expect(wrapper.props('index')).toBe(0);
        expect(wrapper.props('team')).toStrictEqual({id: 0, color: store.state.colorList[0]});
    });

    describe('Input:', () => {
        it('has correct label', () => {
            const wrapper = mount(TeamCreateInput, {
                localVue,
                store,
                propsData: props
            });

            const label = wrapper.find('label[for="team-input"]');
            const color = wrapper.props('team').color.label;

            expect(label.element.innerHTML).toBe(color + ' tým');
        });
    });

    describe("Confirm Button:", () => {
        it('on click commits new team', () => {
            const wrapper = mount(TeamCreateInput, {
                localVue,
                store,
                propsData: props
            });
            const confirmTeamSpy = jest.spyOn(wrapper.vm, "confirmTeam");
            const input = wrapper.find('input');
            const button = wrapper.find('button[name="confirm-button"]');

            expect(store.state.teams.length).toBe(0);
        
            const testLabel = "test";

            input.setValue(testLabel);
            //button.trigger('click');
            button.element.dispatchEvent(new Event('click'))
            expect(confirmTeamSpy).toHaveBeenCalled();
            expect(store.state.teams.length).toBe(1);
            expect(store.state.teams[0]).toStrictEqual({ label: testLabel, color: wrapper.vm.color.id })
        });
    });

    describe('Delete Button', () => {
        it('emits remove() on click', () => {
            const wrapper = mount(TeamCreateInput, {
                localVue,
                store,
                propsData: props
            });
            const button = wrapper.find('button[name="delete-button"]');

            button.trigger('click');

            expect(wrapper.emitted('remove')).toBeTruthy();
        });
    });
});