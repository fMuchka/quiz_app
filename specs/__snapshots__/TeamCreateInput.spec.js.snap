// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamCreateInput Component: matches snapshot 1`] = `
<div
  class="input-wrapper"
>
  <label
    for="team-input"
    style="color: red;"
  >
    Červený tým
  </label>
   
  <input
    name="team-input"
    type="text"
  />
   
  <button
    disabled="disabled"
    name="confirm-button"
  >
    ✓
  </button>
   
  <button
    name="delete-button"
  >
    ✗
  </button>
</div>
`;
